lineax/_solver/lsmr.py

```diff
@@ -0,0 +1,583 @@
+
+            # x is an approximate solution to A@x = b, according to atol and btol.
+            istop += (test1 <= rtol).astype(int) * 2**1
+            # x approximately solves the least-squares problem according to atol.
+            istop += (test2 <= self.atol).astype(int) * 2**2
+            # cond(A) seems to be greater than conlim
+            istop += (test3 <= ctol).astype(int) * 2**3
+
+            # The following tests guard against extremely small values of atol, btol
+            # or ctol.  (The user may have set any or all of the parameters atol, btol,
+            # conlim  to 0.) The effect is equivalent to the normal tests using
+            # atol = eps,  btol = eps,  conlim = 1/eps.
+
+            # x is an approximate solution to A@x = b, according to atol=btol=eps.
+            istop += (1 + t1 <= 1).astype(int) * 2**4
+            # x approximately solves the least-squares problem according to atol=eps.
+            istop += (1 + test2 <= 1).astype(int) * 2**5
+            # cond(A) seems to be greater than 1/eps
+            istop += (1 + test3 <= 1).astype(int) * 2**6
+
+            # maxiter exceeded
+            istop += (itn >= max_steps).astype(int) * 2**7
+
+            return istop == 0
+
+        def bodyfun(loop_state):
+            # unpack everything. Maybe cleaner to use a dict or struct?
+            (
+                loop_state_main,
+                loop_state_r_est,
+                loop_state_anorm,
+                loop_state_stopping,
+                loop_state_vecs,
+            ) = loop_state
+            (
+                itn,
+                alpha,
+                beta,
+                zetabar,
+                alphabar,
+                rho,
+                rhobar,
+                cbar,
+                sbar,
+            ) = loop_state_main
+            (
+                betadd,
+                betad,
+                rhodold,
+                tautildeold,
+                thetatilde,
+                zeta,
+                d,
+            ) = loop_state_r_est
+            (normA2, maxrbar, minrbar, normA, condA) = loop_state_anorm
+            (istop, normr, normb) = loop_state_stopping
+            (x, u, v, h, hbar) = loop_state_vecs
+
+            itn = itn + 1
+
+            # Perform the next step of the bidiagonalization to obtain the
+            # next  beta, u, alpha, v.  These satisfy the relations
+            #         beta*u  =  A@v   -  alpha*u,
+            #        alpha*v  =  A'@u  -  beta*v.
+
+            u = (ω(u) * -alpha).ω
+            u = (ω(u) + ω(operator.mv(v))).ω
+            beta = self.norm(u)
+
+            def beta_gt0(alpha, beta, u, v):
+                u = (ω(u) * (1 / beta)).ω
+                v = (ω(v) * -beta).ω
+                Au = operator.T.mv(u)
+                v = (ω(v) + ω(Au)).ω
+                alpha = self.norm(v)
+                v = lax.cond(alpha > 0, lambda x: (ω(x) / alpha).ω, lambda x: x, v)
```

Comment:
Again I don't really understand this check -- specifically just for the case that `alpha = 0`?

FWIW I think this one has the potential to misbehave. There are circumstances (most notably when vmap'd) under which a `cond` will turn into a `select`, and so both branches are evaluated. In particular that means we will be doing a division-by-zero here, which is probably worth avoiding.

I'd probably suggest a double-where trick:
```
alpha = ...
pred = alpha == 0
safe_alpha = jnp.where(pred, 1 , alpha)
v = jtu.tree_map(lambda vi: jnp.where(pred, 0, vi / safe_alpha), v)
```